corehq/apps/accounting/tasks.py

```diff
@@ -421,6 +422,42 @@ def send_bookkeeper_email(month=None, year=None, emails=None):
         })
 
 
+@periodic_task(run_every=crontab(minute=0, hour=0), acks_late=True)
+def auto_renew_subscriptions():
+    """
+    Automatically renew eligible subscriptions 30 or fewer days from their end date.
+    """
+    ending_subscriptions = _get_auto_renewable_subscriptions()
+    for subscription in ending_subscriptions:
+        if SHOW_AUTO_RENEWAL.enabled(subscription.subscriber.domain) and not subscription.is_renewed:
+            auto_renew_subscription(subscription)
+
+
+def _get_auto_renewable_subscriptions():
+    today = datetime.date.today()
+    date_in_n_days = today + datetime.timedelta(days=30)
+    return Subscription.visible_objects.filter(
+        date_end__gte=today,
+        date_end__lte=date_in_n_days,
+        service_type=SubscriptionType.PRODUCT,
+        auto_renew=True,
+    ).exclude(
+        account__is_customer_billing_account=True,
+    )
+
+
+def auto_renew_subscription(subscription):
+    new_plan_version = DefaultProductPlan.get_default_plan_version(
+        edition=subscription.plan_version.plan.edition,
+        is_annual_plan=subscription.plan_version.plan.is_annual_plan,
+    )
+    subscription.renew_subscription(
+        adjustment_method=SubscriptionAdjustmentMethod.AUTO_RENEWAL,
+        new_version=new_plan_version,
+    )
```

Comment:
Any idea if we need to think about a transaction here? I notice `renew_subscription` is already enclosed in a transaction, so I'm curious if we should be extending that to include fetching the current plan version. Is it an issue if the subscription plan's version were to change between when we initially fetch it and when we renew it?

Reply:
I hadn't considered it, but actually I think the whole thing should be a transaction. If for some reason we fail in notifying the customer, or in (kicking off the task to start) generating and sending an invoice, that should cause a rollback as well. https://github.com/dimagi/commcare-hq/pull/37094/commits/5d58f6777569ce02e66d1936caa00ee1f2c74d3d